server/sonar-server/src/main/java/org/sonar/server/component/index/ComponentIndexer.java

```diff
@@ -64,61 +72,100 @@ public void indexOnStartup(Set<IndexType> uninitializedIndexTypes) {
   }
 
   @Override
-  public void indexProject(String projectUuid, Cause cause) {
+  public void indexOnAnalysis(String projectUuid) {
+    doIndexByProjectUuid(projectUuid, Size.REGULAR);
+  }
+
+  @Override
+  public AuthorizationScope getAuthorizationScope() {
+    return AUTHORIZATION_SCOPE;
+  }
+
+  @Override
+  public Collection<EsQueueDto> prepareForRecovery(DbSession dbSession, Collection<String> projectUuids, Cause cause) {
     switch (cause) {
       case PROJECT_TAGS_UPDATE:
-        break;
+      case PERMISSION_CHANGE:
+        // tags and permissions are not part of type components/component
+        return emptyList();
+
       case PROJECT_CREATION:
+      case PROJECT_DELETION:
       case PROJECT_KEY_UPDATE:
-      case NEW_ANALYSIS:
-        doIndexByProjectUuid(projectUuid, Size.REGULAR);
-        break;
+        List<EsQueueDto> items = projectUuids.stream()
+          .map(projectUuid -> EsQueueDto.create(INDEX_TYPE_COMPONENT.format(), projectUuid, null, projectUuid))
+          .collect(MoreCollectors.toArrayList(projectUuids.size()));
+        return dbClient.esQueueDao().insert(dbSession, items);
+
       default:
         // defensive case
         throw new IllegalStateException("Unsupported cause: " + cause);
     }
   }
 
   @Override
-  public AuthorizationScope getAuthorizationScope() {
-    return AUTHORIZATION_SCOPE;
+  public IndexingResult index(DbSession dbSession, Collection<EsQueueDto> items) {
+    if (items.isEmpty()) {
+      return new IndexingResult();
+    }
+
+    OneToManyResilientIndexingListener listener = new OneToManyResilientIndexingListener(dbClient, dbSession, items);
+    BulkIndexer bulkIndexer = new BulkIndexer(esClient, INDEX_TYPE_COMPONENT, Size.REGULAR, listener);
+    bulkIndexer.start();
+    Set<String> projectUuids = items.stream().map(EsQueueDto::getDocId).collect(MoreCollectors.toHashSet(items.size()));
+    Set<String> remaining = new HashSet<>(projectUuids);
+
+    for (String projectUuid : projectUuids) {
+      // TODO allow scrolling multiple projects at the same time
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)